proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,313 @@
+if the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+After we gain more experience with the restart approach in this proposal, we
+may slowly try to automate more parts to improve cluster reliability.
+
+### Use gossip and consensus to figure out restart slot before the restart
+
+The main difference between this and the current proposal is this alternative
+tries to make the cluster automatically enter restart preparation phase without 
+human intervention.
+
+While getting human out of the loop improves recovery speed, there are concerns 
+about recovery gossip messages interfering with normal gossip messages, and 
+automatically start a new message in gossip seems risky.
+
+### Automatically reduce block production in an outage
+
+Right now we have vote-only mode, a validator will only pack vote transactions
+into new blocks if the tower distance (last_vote - local_root) is greater than
+400 slots.
+
+Unfortunately in the previous outages vote-only mode isn't enough to save the
+cluster. There are proposals of more aggressive block production reduction to
+save the cluster. For example, a leader could produce only one block in four
+consecutive slots allocated to it.
+
+However, this only solves the problem in specific type of outage, and it seems
+risky to aggressively reduce block production, so we are not proceeding with
+this proposal for now.
+
+## Detailed Design
+
+The new protocol tries to make all restarting validators get the same data
+blocks and the same set of last votes, then they will almost certainly make the 
+same decision on the canonical fork and proceed.
+
+A new command line arg will be added. When the cluster is in need
+of a restart, we assume validators holding at least `RESTART_STAKE_THRESHOLD`
+percentage of stakes will restart with this arg. Then the following steps
+will happen:
+
+1. The operator restarts the validator with a new command-line argument to
+cause it to enter the silent repair phase at boot, where it will not make new 
+blocks or vote. The validator propagates its local voted fork
+information to all other validators in restart.
+
+2. While counting local vote information from all others in restart, the
+validator repairs all blocks which could potentially have been optimistically
+confirmed.
+
+3. After repair is complete, the validator counts votes on each fork and
+sends out local heaviest fork.
+
+4. Each validator counts if enough nodes can agree on one block (same slot and
+hash) to restart from:
+
+   1. If yes, proceed and restart
+
+   2. If no, print out what it thinks is wrong, halt and wait for human
+
+See each step explained in details below.
+
+### 1. Gossip last vote before the restart and ancestors on that fork
+
+The main goal of this step is to propagate the locally selected fork to all
```

Comment:
I think we can be more specific, 
locally selected fork" -> 
"the last `n` ancestors of the last voted fork"

Reply:
Changed